### PR TITLE
Zonal face weights at constant lat

### LINKED ISSUE #544: Calculate the weight for each face on the constant latitude
**Describe the feature you'd like added to this project**
The function calculates the weights of grid faces at a specific constant latitude, facilitating accurate non-conservative zonal averages in climate and oceanographic data analysis. This addition aims to optimize and simplify user workflows within our library.

**Link to original functionality implementation elsewhere, if applicable**

**Links to other examples, if applicable**

**Describe the solution you'd like**

- Take a constant latitude as input and compute the perimeter of the corresponding circle.
- Identify the intervals of longitude that overlap with each face of the grid.
- Calculate the weight of each face based on these intervals and the total perimeter.
- Handle any errors or edge cases gracefully, ensuring that the function is robust and user-friendly.

**Additional context**

### PR DESCRIPTION
**Pull request by hongyuchen1030**
Closes #544 

## Overview

      
- The `_get_zonal_face_weights_at_constlat` method has been added to calculate the weight of each face at a constant latitude.
- The user can specify if the given edge is a great circle arc or a constant latitude one
- This method computes the perimeter of a circle at constant latitude, identifies overlapping intervals, and calculates the weight of each face based on the longitude bounds.
- Sweepline algorithm is used to calculate the overlapped intervals' weights
- Errors and edge cases are handled gracefully, with informative error messages and comments for future maintenance.

## Expected Usage

```Python
# Assuming uxarray is imported as ux
grid_path = "/path/to/grid.nc"
data_path = "/path/to/data.nc"

uxds = ux.open_dataset(grid_path, data_path)

# Replace `latitude_rad` with the desired latitude in radians
latitude_rad = 0.0  # Example value

        weight_df = _get_zonal_faces_weight_at_constLat(np.array([
            face_0_edge_nodes, face_1_edge_nodes, face_2_edge_nodes,
            face_3_edge_nodes
        ]),
                                                        0.01,
                                                        latlon_bounds,
                                                        is_directed=False)


```

## PR Checklist

**General**
- [x] An issue is linked created and linked
- [x] Add appropriate labels
- [x] Filled out Overview and Expected Usage (if applicable) sections

**Testing**
- [x] Adequate tests are created if there is new functionality
- [x] Tests cover all possible logical paths in your function
- [x] Tests are not too basic (such as simply calling a function and nothing else)

**Documentation**
- [x] Docstrings have been added to all new functions
- [x] Docstrings have updated with any function changes
- [x] Internal functions have a preceding underscore (`_`) and have been added to `docs/internal_api/index.rst`
- [x] User functions have been added to `docs/user_api/index.rst`

### CODE REVIEW

**Comment by hongyuchen1030** [comment]
@philipc2 I have encountered the failure of the CI saying

```
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_interval - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_GCA_constLat - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
================ 5 failed, 149 passed, 1164 warnings in 47.34s =================
Error: Process completed with exit code 1.
```

However, when I run on my local, all tests passed. My local pandas version is 1.5.2. I wonder if you have any clue about how to resolve this? Thank you very much!

**Comment by philipc2** [comment]
> @philipc2 I have encountered the failure of the CI saying
> 
> ```
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_interval - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_GCA_constLat - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> ================ 5 failed, 149 passed, 1164 warnings in 47.34s =================
> Error: Process completed with exit code 1.
> ```
> 
> However, when I run on my local, all tests passed. My local pandas version is 1.5.2. I wonder if you have any clue about how to resolve this? Thank you very much!

Interesting that it's only passing on MacOS. I don't believe it's a pandas version issue, but I would definitely update it to match with the version in the environment used for our CI tests.  

I gave the code a quick look through and couldn't figure out where that issue is occurring. I can give it a look later on.

**Comment by hongyuchen1030** [comment]
> > @philipc2 I have encountered the failure of the CI saying
> > ```
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_interval - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_GCA_constLat - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > ================ 5 failed, 149 passed, 1164 warnings in 47.34s =================
> > Error: Process completed with exit code 1.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, when I run on my local, all tests passed. My local pandas version is 1.5.2. I wonder if you have any clue about how to resolve this? Thank you very much!
> 
> Interesting that it's only passing on MacOS. I don't believe it's a pandas version issue, but I would definitely update it to match with the version in the environment used for our CI tests.
> 
> I gave the code a quick look through and couldn't figure out where that issue is occurring. I can give it a look later on.

Thank you very much for your help. As you can see, I have removed the `pandas.dataframe.append()` from my codes, and the current error messages on other system are that they can't find the intersection points, which is really abnormal since it's passing on MacOS. 

```
=========================== short test summary info ============================
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - ValueError: UXarray doesn't support concave face with intersections points as [3] currently, please modify your grids accordingly
```

**Comment by hongyuchen1030** [comment]
Note: Instead of using the `is_GCA_list`, use the flag as `is_unstructured`. If the `is_unstructured == true`, treat edge that has two endpoints that have the same latitude as the constant latitude

**Comment by hongyuchen1030** [comment]
Now the `is_latlonface` flag is added, which will be applied to all grids. But the `is_GCA` flag is still preserved and it can overwrite all other specification of the edge type.

Important note: The function will only accept convex face as input now. A more general version that take cares of the concave face will be implemented in the future. That means even if a polygon looks convex on a plane, it might look concave on the sphere, and our current function will still throw an exception for that

File: uxarray/grid/integrate.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+        if (
+            np.isclose(n1[2], latZ, rtol=0, atol=ERROR_TOLERANCE)
+            and np.isclose(n2[2], latZ, rtol=0, atol=ERROR_TOLERANCE)
+            and is_GCA is False
+        ):
+            overlap_flag = True
+
+        # Skip the dummy edge
+        if np.any(n1 == [INT_FILL_VALUE, INT_FILL_VALUE, INT_FILL_VALUE]) or np.any(
+            n2 == [INT_FILL_VALUE, INT_FILL_VALUE, INT_FILL_VALUE]
```
Comment (philipc2) [inline uxarray/grid/integrate.py:175]: Can you turn `[INT_FILL_VALUE, INT_FILL_VALUE, INT_FILL_VALUE]` into a variable for readability?

File: uxarray/grid/integrate.py [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+        face_index: overlap_contributions.get(face_index, 0.0) / total_length
+        for face_index in range(len(faces_edges_cart))
+    }
+
+    # Convert weights to DataFrame
+    weights_df = pd.DataFrame(list(weights.items()), columns=["face_index", "weight"])
+    return weights_df
+
+
+def _get_zonal_face_interval(
```
Comment (philipc2) [inline uxarray/grid/integrate.py:92]: This function is extremely long, with multiple if-else blocks that could be turned into helper functions. Can you turn some of those if-else blocks into helper functions?

File: uxarray/grid/integrate.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
+"""uxarray grid module."""
```
Comment (philipc2) [inline uxarray/grid/integrate.py:1]:
```suggestion
```

File: uxarray/grid/integrate.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+
```
Comment (philipc2) [inline uxarray/grid/integrate.py:2]: I'm not sure if `integrate.py` is the best module name for this.

**Review by philipc2 — CHANGES_REQUESTED** [review]

File: uxarray/grid/integrate.py (hunk 4 above)
Comment (hongyuchen1030) [reply uxarray/grid/integrate.py:1]: Do you have any better suggestion? I call it "integrate" because it is described in our "Integration Algorithm" paper

**Review by erogluorhan — COMMENTED** [review]
Re current CI failures:

With the commit entitled "[WIP](https://github.com/UXARRAY/uxarray/pull/555/commits/1598de2650bd9873482221ba774fe3e2b668be3a)" on the Feb 4th this year, several lines of codes seem to have been added. The following code block:

```
unique_intersection_lonlat = np.array(
                    [node_xyz_to_lonlat_rad(pt.tolist()) for pt in unique_intersection]
                )
>               unique_intersection_lonlat = unique_intersection_lonlat.append(
                    [2 * np.pi, latitude_rad]
                )
```

apparently creates a `numpy.ndarray` then attempts to `.append()` into that, which results in 

`AttributeError: 'numpy.ndarray' object has no attribute 'append'`

**Comment by hongyuchen1030** [comment]
> Re current CI failures:
> 
> With the commit entitled "[WIP](https://github.com/UXARRAY/uxarray/pull/555/commits/1598de2650bd9873482221ba774fe3e2b668be3a)" on the Feb 4th this year, several lines of codes seem to have been added. The following code block:
> 
> ```
> unique_intersection_lonlat = np.array(
>                     [node_xyz_to_lonlat_rad(pt.tolist()) for pt in unique_intersection]
>                 )
> >               unique_intersection_lonlat = unique_intersection_lonlat.append(
>                     [2 * np.pi, latitude_rad]
>                 )
> ```
> 
> apparently creates a `numpy.ndarray` then attempts to `.append()` into that, which results in
> 
> `AttributeError: 'numpy.ndarray' object has no attribute 'append'`
> 
> FYI: The last successful CI actions before the above commit seems to be [Merge branch 'main' into zonal_face_weights_at_constantlat](https://github.com/UXARRAY/uxarray/pull/555/commits/79bf078a90b47c2fa270ae20c75de70a7842ea31), which was on Jan 16.

Thanks for your information. Yeah I also noticed that this was the cause. So this syntax is actually valid before. I assume it was the version upgrade that makes it corrupts. I am still working on fixing it since I have been working on other research project recently.

**Comment by erogluorhan** [comment]
> Thanks for your information. Yeah I also noticed that this was the cause. So this syntax is actually valid before. I assume it was the version upgrade that makes it corrupts. I am still working on fixing it since I have been working on other research project recently.

I think you intended to use `numpy.append(array-like1, array-like2,...)` rather than the one here (i.e. non-existent `numpy.ndarray.append()`?

**Comment by hongyuchen1030** [comment]
> > Thanks for your information. Yeah I also noticed that this was the cause. So this syntax is actually valid before. I assume it was the version upgrade that makes it corrupts. I am still working on fixing it since I have been working on other research project recently.
> 
> I think you intended to use `numpy.append(array-like1, array-like2,...)` rather than the one here (i.e. non-existent `numpy.ndarray.append()`?

I guessed the same thing before. But when I tried to fix it by using this `numpy.append(array-like1, array-like2,...)` , all my following codes and functions call corrupts due to incorrect array shape, which is not logical since my previous implementation passed all the test cases. I am still trying to figure out what I intended to do for that WIP commit. I had been actively working on this branch to test many current research ideas since it's been hanging there for a while.

**Comment by hongyuchen1030** [comment]
Hi @philipc2 @erogluorhan,

I have cleaned up and fixed the CI for this branch and it is mostly ready to go now. However, the read-doc failed is currently failing because the module `uxarray.Grid.grid_spec.`, which I didn't modify in this PR. I wonder if you can take a look into this? Thanks. Other than that, this PR is ready to be merged

> Running Sphinx v7.3.6
> making output directory... done
> [autosummary] generating autosummary for: api.rst, citation.rst, contributing.rst, examples.rst, examples/000-template.ipynb, examples/001-working-with-unstructured-grids.ipynb, examples/002-grid-topology.ipynb, examples/003-area-calc.ipynb, examples/004-working-with-mpas-grids.ipynb, examples/005-to-geodataframe-for-holoviz.ipynb, ..., examples/009-subsetting.ipynb, getting-started/freq-asked-questions.rst, getting-started/overview.rst, getting-started/quick-install.rst, index.rst, installation.rst, internal_api/index.rst, quickstart.rst, tutorials.rst, user_api/index.rst
> WARNING: [autosummary] failed to import uxarray.Grid.grid_spec.
> Possible hints:
> * PycodeError: no source found for module 'builtins'
> * AttributeError: type object 'Grid' has no attribute 'grid_spec'
> * ModuleNotFoundError: No module named 'uxarray.Grid'
> WARNING: [autosummary] failed to import uxarray.Grid.n_max_node_nodes.
> Possible hints:
> * AttributeError: type object 'Grid' has no attribute 'n_max_node_nodes'
> * PycodeError: no source found for module 'builtins'
> * ModuleNotFoundError: No module named 'uxarray.Grid'
> WARNING: [autosummary] failed to import uxarray.grid._slice_edge_indices.
> Possible hints:
> * ModuleNotFoundError: No module named 'uxarray.grid._slice_edge_indices'
> * ImportError: 
> * AttributeError: module 'uxarray.grid' has no attribute '_slice_edge_indices'
> WARNING: [autosummary] failed to import uxarray.grid._slice_face_indices.
> Possible hints:
> * ImportError: 
> * AttributeError: module 'uxarray.grid' has no attribute '_slice_face_indices'
> * ModuleNotFoundError: No module named 'uxarray.grid._slice_face_indices'
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._build_face_edges_connectivity.
> Possible hints:
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._build_face_edges_connectivity'; 'uxarray.grid.connectivity' is not a package
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_build_face_edges_connectivity'
> * ImportError: 
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._build_nNodes_per_face.
> Possible hints:
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_build_nNodes_per_face'
> * ImportError: 
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._build_nNodes_per_face'; 'uxarray.grid.connectivity' is not a package
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._populate_face_edges_connectivity.
> Possible hints:
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_populate_face_edges_connectivity'
> * ImportError: 
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._populate_face_edges_connectivity'; 'uxarray.grid.connectivity' is not a package
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._populate_node_faces_connectivity.
> Possible hints:
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_populate_node_faces_connectivity'
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._populate_node_faces_connectivity'; 'uxarray.grid.connectivity' is not a package
> * ImportError:

**Comment by philipc2** [comment]
> Details

Hi @hongyuchen1030 

Thanks for the revisions! Giving this a look through now and I'll look into what's causing the RTD failures.

**Comment by hongyuchen1030** [comment]
> > Details
> 
> Hi @hongyuchen1030
> 
> Thanks for the revisions! Giving this a look through now and I'll look into what's causing the RTD failures.

Thank you very much! I just fixed a typo in the .rst file, I think the read-docs works now

**Review by philipc2 — APPROVED** [review]
This looks fantastic! Thanks for addressing all the comments! 

The documentation appears to be building fine. I believes the output you had were just warnings.

**Review by erogluorhan — APPROVED** [review]
This looks good to me. Thanks for addressing all the issues!

**Review by erogluorhan — APPROVED** [review]

**Pull request merged**